packages/lending/src/adapter.ts

```diff
@@ -0,0 +1,966 @@
+    }
+    if (!bestSwaper) {
+      bestSwaper = this.swapers[0];
+    }
+
+    return bestSwaper;
+  }
+
+  async getPortfolios(account: string): Promise<Portfolio[]> {
+    const portfolios = await Promise.all(
+      Object.values(this.protocolMap).map((protocol) => protocol.getPortfolios(account))
+    );
+    return flatten(portfolios);
+  }
+
+  async getPortfolio(account: string, protocolId: string, _marketId: string) {
+    return await this.protocolMap[protocolId].getPortfolio(account, _marketId);
+  }
+
+  getProtocol(id: string) {
+    return this.protocolMap[id];
+  }
+
+  // tokenA: USDC, tokenB: USDT
+  // 1. flashloan USDC
+  // 2. swap USDC to USDT
+  // 3. deposit USDT, get aUSDT
+  // 4. return-funds aUSDT to user
+  // 5. add-funds aUSDC to router
+  // 6. withdraw aUSDC, get USDC
+  // 7. flashloan repay USDC
+  async getCollateralSwap(
+    protocolId: string,
+    marketId: string,
+    params: BaseParams,
+    account: string,
+    portfolio?: Portfolio
+  ): Promise<BaseFields> {
+    const { srcAmount } = params;
+    const srcToken = common.classifying(params.srcToken);
+    const destToken = common.classifying(params.destToken);
+    const wrappedSrcToken = wrapToken(this.chainId, srcToken);
+    const wrappedDestToken = wrapToken(this.chainId, destToken);
+    const collateralSwapLogics: api.Logic<any>[] = [];
+    const protocol = this.getProtocol(protocolId);
+
+    portfolio = portfolio || (await protocol.getPortfolio(account, marketId));
+    const afterPortfolio = portfolio.clone();
+
+    // ---------- flashloan ----------
+    const flashLoanAggregatorQuotation = await api.protocols.utility.getFlashLoanAggregatorQuotation(this.chainId, {
+      repays: [{ token: wrappedSrcToken, amount: srcAmount }],
+      protocolId: protocolId,
+    });
+
+    const flashLoanTokenAmount = flashLoanAggregatorQuotation.loans.tokenAmountMap[wrappedSrcToken.address];
+
+    const [flashLoanLoanLogic, flashLoanRepayLogic] = api.protocols.utility.newFlashLoanAggregatorLogicPair(
+      flashLoanAggregatorQuotation.protocolId,
+      flashLoanAggregatorQuotation.loans.toArray()
+    );
+    collateralSwapLogics.push(flashLoanLoanLogic);
+
+    // ---------- swap ----------
+    const swaper = this.findSwaper([wrappedSrcToken, wrappedDestToken]);
+    const swapQuotation = await swaper.quote({
+      input: flashLoanTokenAmount,
+      tokenOut: wrappedDestToken,
+      slippage: defaultSlippage,
+    });
+
+    const swapTokenLogic = swaper.newSwapTokenLogic(swapQuotation);
+    collateralSwapLogics.push(swapTokenLogic);
+
+    // ---------- supply ----------
+    const supplyLogic = await protocol.newSupplyLogic({ input: swapQuotation.output, marketId });
+
+    collateralSwapLogics.push(supplyLogic);
+
+    afterPortfolio.supply(swapQuotation.output.token, swapQuotation.output.amount);
+
+    // compound v3 base token supplied not support collateral swap
+    if (protocolId === 'compoundv3' && supplyLogic.fields.output)
```

Comment:
Would it be better to check if supply token equals base token?
Also suggest doing this check at the beginning to save resources.

Reply:
Add baseToken to `Portfolio` so that we can use it easily, and move this condition to save resources.

```typescript
// compound v3 base token supplied not support collateral swap
    if (protocolId === 'compound-v3' && portfolio.baseToken?.is(srcToken.unwrapped))
      throw new Error('Compound V3 does not support the base token for performing collateral swaps');
```